src/build_system.ml

```diff
@@ -323,21 +325,192 @@ module Dir_status = struct
     ; loc : Loc.t option
     }
 
+  module Alias : sig
+    type t
+    type frozen
 
-  type alias =
-    { mutable deps     : Path.Set.t
-    ; mutable dyn_deps : (unit, Path.Set.t) Build.t
-    ; mutable actions  : alias_action list
-    }
+    val create : unit -> t
 
-  type rules_collector =
-    { mutable rules   : Build_interpret.Rule.t list
-    ; mutable aliases : alias String.Map.t
-    ; mutable stage   : collection_stage
-    }
+    type immutable =
+      { deps     : Path.Set.t
+      ; dyn_deps : (unit, Path.Set.t) Build.t
+      ; actions  : alias_action list
+      }
+
+    val _of_immutable : immutable -> frozen
```

Comment:
Why do we need this function? Is there a future commit that requires it?

Reply:
Related: do we need the two `frozen` and `immutable` types? It seems to me that they could be the same type

Reply:
Good points. Removed both. (background: I started with just the "frozen" type and later needed "immutable" because call sites use the record fields, but I failed to realize that "frozen" type is not necessary then)